File: android/crosspl/anno/src/main/kotlin/org/elastos/tools/crosspl/processor/generator/CrossPLUtilsGenerator.kt
package org.elastos.tools.crosspl.processor.generator

import org.elastos.tools.crosspl.processor.CrossTmplUtils
import org.elastos.tools.crosspl.processor.Log
import java.io.File

class CrossPLUtilsGenerator {
    companion object {
        fun Generate(crossplDir: File): Boolean {
            val headerFile = GetHeaderFile(crossplDir)
            val sourceFile = GetSourceFile(crossplDir)

            var ret = GenerateFile(CrossPLUtilsHeaderTmpl, headerFile)
            if(! ret) {
                return ret
            }

            ret = GenerateFile(CrossPLUtilsSourceTmpl, sourceFile)
            if(! ret) {
                return ret
            }

            var spanFile = GetFile(crossplDir, ExperimentalSpanTmpl)
            ret = GenerateFile(ExperimentalSpanTmpl, spanFile)
            if(! ret) {
                return ret
            }

            return true
        }

        fun GetSourceFile(crossplDir: File): File {
            return GetFile(crossplDir, CrossPLUtilsSourceTmpl)
        }

        fun GetHeaderFile(crossplDir: File): File {
            return GetFile(crossplDir, CrossPLUtilsHeaderTmpl)
        }

        private fun GenerateFile(from: String, to: File): Boolean {
            Log.w("Generate: ${to.absolutePath}")
            val content = CrossTmplUtils.ReadTmplContent(from)

            CrossTmplUtils.WriteContent(to, content)
            return true
        }

        fun GetFile(crossplDir: File, fileName: String): File {
            return File(crossplDir, fileName)
        }

        private const val CrossPLUtilsHeaderTmpl = "/CrossPLUtils.hpp"
        private const val CrossPLUtilsSourceTmpl = "/CrossPLUtils.cpp"
        private const val ExperimentalSpanTmpl = "/experimental-span.hpp"
    }
}

File: ios/anno/swift/generator/CrossPLUtilsGenerator.swift
import Foundation

class CrossPLUtilsGenerator {
  static func Generate(crossplDir: URL) -> Bool {
    let headerFile = GetHeaderFile(crossplDir: crossplDir)
    let sourceFile = GetSourceFile(crossplDir: crossplDir)
  
    var ret = GenerateFile(CrossPLUtilsHeaderTmpl, headerFile)
    if ret == false {
      return ret
    }
  
    ret = GenerateFile(CrossPLUtilsSourceTmpl, sourceFile)
    if ret == false {
      return ret
    }
  
    let spanFile = GetFile(crossplDir, ExperimentalSpanTmpl)
    ret = GenerateFile(ExperimentalSpanTmpl, spanFile)
    if ret == false {
      return ret
    }
  
    return true
  }
  
  static func GetSourceFile(crossplDir: URL) -> URL {
    return GetFile(crossplDir, CrossPLUtilsSourceTmpl)
  }
  
  static func GetHeaderFile(crossplDir: URL) -> URL {
    return GetFile(crossplDir, CrossPLUtilsHeaderTmpl)
  }
  
  private static func GenerateFile(_ from: String, _ to: URL) -> Bool {
    print("Generate: \(to.path)")
    let content = CrossTmplUtils.ReadTmplContent(tmplName: from)
  
    CrossTmplUtils.WriteContent(file: to, content: content)
    return true
  }
  
  static func GetFile(_ crossplDir: URL, _ fileName: String) -> URL {
    return crossplDir.appendingPathComponent(fileName)
  }
  
  private static let CrossPLUtilsHeaderTmpl = "/CrossPLUtils.h"
  private static let CrossPLUtilsSourceTmpl = "/CrossPLUtils.mm"
  private static let ExperimentalSpanTmpl = "/experimental-span.hpp"
}
